### ISSUE TITLE
ValueObservation notifies duplicate initial value when started from a DatabasePool

### ISSUE THREAD

**Issue by pocketpixels** [groue/GRDB.swift#937]
### What did you do?
I just updated from GRDB 4 + GRDBCombine to GRDB 5. 
It was surprisingly painless and the only code change I had to make was to replace the `fetchOnSubscription()` calls with `scheduling: .immediate` when creating the Combine publishers. 

### What did you expect to happen?

I expected identical notification behavior as before. That is, one initial value notification right away, and then one notification for each observed value change after that. 

### What happened instead?

The initial value is being sent twice. 

Dealing with the duplicate notification is trivial. 
However the fact that the SQL query is executed twice can be a potentially significant performance regression.
(In my application this is an issue on the watch)

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.5.0
**Installation method:** CocoaPods
**Xcode version:** 12.5 beta 3
**Swift version:** 5
**Platform(s) running GRDB:** iOS and watchOS
**macOS version running Xcode:**

**Comment by groue** [comment]
Hello @pocketpixels,

This is a known behavior of the GRDB 5 DatabasePool that I'm still struggling to avoid.

It is not strictly a bug, because the documentation carefully says that ValueObservation can notify duplicates: apps should be ready for such stuttering.

But just as you say, this can create performance issues. And this is just weird. And once you notice it, you can't forget it. Meh.

There is a benefit, though. When you start a ValueObservation on a DatabasePool, you are now guaranteed to get the initial value shortly, even if there is a long write transaction in the background. With GRDB 4, observations could only start *after* all pending writes were completed, and this could create a really bad user experience. Slow writes happen, for example, in apps that sync a lot of data with a remote server, and those *could not* display data on screen until those long writes would complete (#601, addressed by #736).

So for one win (non-blocking observations), we get one loss (double notification of the initial value).

There is already a fix, for custom SQLite builds compiled with the `SQLITE_ENABLE_SNAPSHOT` option. GRDB can then avoid the double initial notification. This option is unfortunately not available on the stock SQLite version that ships with iOS and watchOS. Well, it's there, but unusable, as cruelly revealed by #846.

The difficulty is that when you start a `ValueObservation`:
1. An initial value is fetched as soon as possible from a reader SQLite connection
2. The observation then waits for a slot in the writer SQLite connection
3. From the writer SQLite connection, it can hook on SQLite commits and spot changes that impact the observed region. From now on, no change will be missed.
4. Between 1 and 3, the database could have been modified, and nobody was aware of it.
5. Hence the security double fetch, just in case we missed a change.

The difficulty is (4): if we can prove for sure that between the initial read in a reader connection (1), and the setup of the commit hooks on the writer connection (3), no change was performed, then we can avoid the double notification.

It's possible with `SQLITE_ENABLE_SNAPSHOT`.

Without it, I could never manage to do it. Some serious SQLite mojo is needed here 😅

**Comment by groue** [comment]
> (In my application this is an issue on the watch)

I currently have no real workaround. You can try `DatabaseQueue`, which does not suffer from this problem.

**Comment by pocketpixels** [comment]
Hi Gwendal,
thank you very much for the detailed reply and discussion. Much appreciated! 

Could it be possible to start with step 2, check if the writer connection is idle (with nothing queued up) and if yes, synchronously execute the initial fetch on that (doing the check and scheduling as one atomic operation)? And perform step 1 only if the writer connection is not idle?

**Comment by pocketpixels** [comment]
Switching to the `DatabaseQueue` is probably the way to go for my application. 
I sync with CloudKit and the initial sync can be large and slow. However I am not keeping a GRDB write transaction open while downloading info from CloudKit, I only write a batch of changes to the database after completing the Cloudkit fetch for it. So the writes should be quick and reads being temporarily blocked should not be a major issue.

**Comment by groue** [comment]
> Could it be possible to start with step 2, check if the writer connection is idle (with nothing queued up)

There's no such API in DispatchQueue.

It may well be possible, when the ValueObservation is started asynchronous (not `.immediate`), to start a race between the writer queue and a reader queue, in order to have a chance to start straight from the writer queue (and thus avoid the double initial fetch). But there is no way to tell the race to prefer one side, and we'd end up with a quite unpredictable behavior.

Generally speaking, I think observations that perform their initial fetch from the writer queue are off the table. Some users would complain that such an observation postpones writes, in a noticeable fashion when the fetch is slow. The *purpose* of `DatabasePool` is to allow parallel reads and writes, after all.

So I'm still looking at another technique.

> Switching to the DatabaseQueue is probably the way to go for my application.

Yes, please give it a try.

Nevertheless, I hope we can make progress on the topic.

A few days ago I stumbled upon [SQLITE_FCNTL_DATA_VERSION](https://www.sqlite.org/c3ref/c_fcntl_begin_atomic_write.html#sqlitefcntldataversion) and had great hopes. Unfortunately, this would not give the expected result.

The **great difficulty** is to tell if the database is different when seen from one SQLite connection (the reader that performs the initial fetch), and another SQLite connection (the writer where observation really starts). The `SQLITE_ENABLE_SNAPSHOT` compile-time option enables [sqlite3_snapshot_cmp](https://www.sqlite.org/c3ref/snapshot_cmp.html), whose *purpose* is precisely to compare connection-independent databases snapshots. This is the perfect, SQLite-level solution, as reliable as one can dream.

Maybe, as you suggest, we can acquire the same level of robustness with DispatchQueue juggling at the GRDB level. Good ideas are always welcome!

**Comment by groue** [comment]
I'll keep this issue open until we find a fix.

**Comment by groue** [comment]
> I'll keep this issue open until we find a fix. It's not strictly a bug, but it's a real inconvenience.

I just can't leave issues open, especially during a lockdown: #940

**Comment by groue** [comment]
This issue can be solved with a [custom SQLite build](https://github.com/groue/GRDB.swift/blob/master/Documentation/CustomSQLiteBuilds.md), with the `SQLITE_ENABLE_SNAPSHOT` flag.

I reported to Apple the need for these apis (FB9793771 - Expose SQLite APIs enabled by the SQLITE_ENABLE_SNAPSHOT compilation option).

I'm not aware of any other solution. When a ValueObservation fetches its initial value without waiting for a write access, we _need_ a way to tell if the initial read was performed on the same database state than the one found when the transaction observer is installed. `SQLITE_ENABLE_SNAPSHOT` makes it possible to answer this question. Without this flag, ValueObservation must perform a second fetch, just in case database would have been changed.

I'm closing this issue with the "needs revisiting" flag, because there's not much that can be done at this point.
